New: "How WinRT types and members are represented in JavaScript" (#2343)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/how-to/winrt-js-conversion.md

---
title: How WinRT types and members are represented in JavaScript
description: How to use WinRT types and members from within web-side JavaScript code in a WebView2 app, when calling native-side WinRT code from web-side code.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
ms.date: 12/02/2022

> ---

Author: global - delete "instance"?  it might just be introducing confusion

Author: no change, so far.  asked in chat

| `Enum` | JavaScript object | An enum type is represented as a JavaScript object.  Each enum value is a `Number` property on the JavaScript object. |
| `Struct` | JavaScript object | A `Struct` type is converted to a JavaScript object that has property names which correspond to the `Struct` type member names.  This is a two-way conversion. |
| `Namespace` | JavaScript object | A namespace is represented as a JavaScript object that has a property for each child namespace, enum type, and `RuntimeClass`. |
| `Class` | JavaScript object | A `RuntimeClass` class is converted to a JavaScript object that has the same methods, properties, and events. |
| `Interface` | JavaScript object | A `RuntimeClass` interface is converted to a JavaScript object that has the same methods, properties, and events.  There is no support for implementing an interface in JavaScript. |
| Class static member | JavaScript object property | See below. |
| Class constructor | JavaScript constructor and function | See below. |

See also:
* [RuntimeClass class](https://learn.microsoft.com/en-us/cpp/cppcx/wrl/runtimeclass-class)

<!-- ------------------------------ -->
#### Class static members

> <!-- orig wording:

Author: @nishitha-burman - pls tech review the rewrite of this sentence below

Reviewer: @david-risney for tech review

Author: See the additional draft wording phrases at latest Files Changed tab:
https://github.com/MicrosoftDocs/edge-developer/pull/2343/files
Find "Class static members"

Author: probably covered by verbose explanation added to sibling section

Runtimeclasses with static methods, properties, or events, are represented as properties of their namespace with the static methods, properties, and events as properties on that JavaScript object.
-->
A `RuntimeClass` class that has static methods, properties, or events is represented as a property of the corresponding JavaScript namespace object.  The static methods, properties, and events are represented as properties on the JavaScript namespace object.

For example, to call the static method `Windows.Foundation.Uri.EscapeComponent`, call:

```javascript
`chrome.webview.hostObjects.Windows.Foundation.Uri.EscapeComponent("example");`
```

The JavaScript namespace object here is `chrome.webview.hostObjects`.

<!-- ------------------------------ -->
#### Class constructors

> <!-- orig wording:

Author: 2 - @nishitha-burman - pls tech review the rewrite of this sentence below

Author: probably handled by other cmt below

| `IVector<T>`, `IVectorView<T>`, `IObservableVector<T>` | `Array` and JavaScript object | If a `RuntimeClass` instance implements `vector` interfaces, then it is represented in JavaScript as the usual object described below, but also will act like a JavaScript array.  Reads and writes are performed live on the underlying WinRT vector object.
| `IMap<K,V>`, `IMapView<K,V>`, `IObservableMap<K,V>` | JavaScript object | If a `RuntimeClass` instance implements `map` interfaces, then it is represented in JavaScript as the usual object described below, but also has properties with name and values from the underlying WinRT map object.  Reads and writes are performed live on the underlying WinRT map object. |
| `Enum` | JavaScript object | An enum type is represented as a JavaScript object.  Each enum value is a `Number` property on the JavaScript object. |
| `Struct` | JavaScript object | A `Struct` type is converted to a JavaScript object that has property names which correspond to the `Struct` type member names.  This is a two-way conversion. |

> | `Namespace` | JavaScript object | A namespace is represented as a JavaScript object that has a property for each child namespace, enum type, and `RuntimeClass`. |

Reviewer 2: It feels like 'child namespace, enum type, and runtimeclass" should all be plural. A namespace may contain none or some child namespaces, child enum types, and child runtimeclass types.

```suggestion
| `Namespace` | JavaScript object | A namespace is represented as a JavaScript object that has a property for each child namespace, enum type, and `RuntimeClass`. |
```

Author: ```
| `Namespace` | JavaScript object | A namespace is represented as a JavaScript object that has a property for any child namespaces, enum types, or `RuntimeClass`.  The namespace may have 0, 1, or many child namespaces, enums, or runtimeclasses, and each individual child namespace, enum and runtimeclass gets its own property. |
```

| `Class` | JavaScript object | A `RuntimeClass` class is converted to a JavaScript object that has the same methods, properties, and events. |
| `Interface` | JavaScript object | A `RuntimeClass` interface is converted to a JavaScript object that has the same methods, properties, and events.  There is no support for implementing an interface in JavaScript. |
| Class static member | JavaScript object property | See below. |
| Class constructor | JavaScript constructor and function | See below. |

See also:

> * [RuntimeClass class](https://learn.microsoft.com/en-us/cpp/cppcx/wrl/runtimeclass-class)

Reviewer 2: We shouldn't link to c++/cx. Microsoft is recommending c++/winrt now instead. A better link might be https://learn.microsoft.com/en-us/uwp/midl-3/intro

Author: `* [Introduction to Microsoft Interface Definition Language 3.0](/uwp/midl-3/intro)`

<!-- ------------------------------ -->
#### Class static members

<!--
original wording:
Runtimeclasses with static methods, properties, or events, are represented as properties of their namespace with the static methods, properties, and events as properties on that JavaScript object.

that might mean:

The runtime class that has static properties, methods, or events is a property of the namespace.  Each static property, method, or event is a property of the JavaScript Object.

A `RuntimeClass` class that has static properties, methods, or events is represented as a property of the corresponding JavaScript namespace object.  The static methods, properties, and events are represented as properties on the JavaScript namespace object.
-->

> A runtime class that has static properties, static methods, or static events is represented as a property of the namespace.  Each static property, method, or event is represented as a property on that JavaScript object.

Reviewer 2: Sorry for my confusing wording. Here are some potential updates to the wording and then trying to explain what I mean more below

```suggestion
A runtime class that has static properties, static methods, or static events is represented as a property of the namespace.  Each static property, static method, and static event is represented as a property on that JavaScript object of the runtimeclass.
```

Eg for the WinRT API static method `Windows.Foundation.Uri.EscapeComponent`, `Windows.Foundation` is the namespace, `Uri` is the runtimeclass, and `EscapeComponent` is the static method. In JavaScript it looks similar: `chrome.webview.hostObjects.Windows.Foundation.Uri.EscapeComponent` where the EscapeComponent is a JS method that is a property on the JS object for the Uri runtimeclass, which is a property on the JS object for the Foundation namespace.

Author: snapshot of result i'm trying:
```
A runtime class that has static properties, static methods, or static events is represented as a property of the namespace. Each static property, static method, and static event is represented as a property on that JavaScript object of the runtimeclass.

For example, for the WinRT API static method `Windows.Foundation.Uri.EscapeComponent`:
*  `Windows.Foundation` is the namespace.
*  `Uri` is the runtimeclass.
*  `EscapeComponent` is the static method.

In JavaScript, the representation looks similar: `chrome.webview.hostObjects.Windows.Foundation.Uri.EscapeComponent`:
*  `EscapeComponent` is a JavaScript method that's a property on the JavaScript object for the `Uri` runtimeclass.
*  The `Uri` runtimeclass is a property on the JavaScript object for the `Foundation` namespace.
```

For example, to call the static method `Windows.Foundation.Uri.EscapeComponent`, call:

```javascript
`chrome.webview.hostObjects.Windows.Foundation.Uri.EscapeComponent("example");`
```

> The JavaScript namespace object here is `chrome.webview.hostObjects`.

Author: @david-risney - should this be (like done below):
The JavaScript namespace object here is `chrome.webview.hostObjects.Windows.Foundation`.

Reviewer 2: Yes, thx

Author: fixed locally

<!-- ------------------------------ -->
#### Class constructors

<!-- orig wording:
Runtimeclasses with constructors are represented as a JavaScript constructor and function on the namespace object.
-->

> A constructor for a `RuntimeClass` class is represented as two members on a JavaScript object:

Reviewer 2: There's just one property, one thing in JS but it can be called two different ways, either as a function or with new as a constructor:

```suggestion
A constructor for a `RuntimeClass` class is represented as a single property on a JavaScript object that can be called two ways:
```

Author: fixed locally

> *  A constructor on the JavaScript namespace object.

Reviewer 2: ```suggestion
*  As a constructor on the JavaScript namespace object.
```

Author: fixed both list items locally

*  A function on the JavaScript namespace object.

For example, to create a new `Windows.Foundation.Uri` object, you can either call it as a constructor, using `new`:

```javascript
`let uri = new chrome.webview.hostObjects.Windows.Foundation.Uri("https://example.com/");`
```

Or, call it as a function, without `new`:

```javascript
`let uri = chrome.webview.hostObjects.Windows.Foundation.Uri("https://example.com/");`
```

> The JavaScript namespace object here is `chrome.webview.hostObjects`.

Reviewer 2: This is a slightly confusing statement. The WinRT namespace objects would be the Windows and Foundation objects. 'chrome.webview.hostObjects' might be thought of as a sort of namespace which contains all of the stuff we're talking about here, but confusing when we are talking about JS objects that represent actual WinRT namespaces

```suggestion
The JavaScript namespace object here is `chrome.webview.hostObjects.Windows.Foundation`.
```

Author: fixed locally (2x)
